src/librustc_typeck/astconv.rs

```diff
@@ -1487,14 +1619,37 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx> + 'o {
                 self.prohibit_generics(path.segments.split_last().unwrap().1);
                 self.ast_path_to_ty(span, did, path.segments.last().unwrap())
             }
-            Def::Variant(did) if permit_variants => {
+            Def::Variant(_) if permit_variants => {
                 // Convert "variant type" as if it were a real type.
                 // The resulting `Ty` is type of the variant's enum for now.
                 assert_eq!(opt_self_ty, None);
-                self.prohibit_generics(path.segments.split_last().unwrap().1);
-                self.ast_path_to_ty(span,
-                                    tcx.parent_def_id(did).unwrap(),
-                                    path.segments.last().unwrap())
+
+                let path_segs = self.def_ids_for_path_segments(&path.segments, None, path.def);
+                let generic_segs: FxHashSet<_> =
+                    path_segs.iter().map(|PathSeg(_, index)| index).collect();
+                self.prohibit_generics(path.segments.iter().enumerate().filter_map(|(index, seg)| {
+                    if !generic_segs.contains(&index) {
+                        Some(seg)
+                    } else {
+                        None
+                    }
+                }));
+
+                let PathSeg(def_id, index) = path_segs.last().unwrap();
+                self.ast_path_to_ty(span, *def_id, &path.segments[*index])
```

Comment:
@petrochenkov Do we need a `check_stability` call on the variant here, like in `associated_path_def_to_ty`? (Note, there wasn't one before my change here even.)

Reply:
Yep... how come it isn't needed in `def_to_ty` too, out of curiosity?

Reply:
Variant in `def_to_ty` is not an associated item and is resolved early, so it's checked in the primary stability pass in `stability.rs`.

Reply:
Actually, why do you need this code in `def_to_ty` at all?
It should only affect variant paths not involving aliases.

Reply:
Aha, I see.

Anyway, I need this to consistently support `Enum::<...>::Variant`, that's all – since we're supporting it elsewhere now. :-)